### PR TITLE
Reg arbitration

### PR DESCRIPTION
**Pull request by tastynoob**

### CODE REVIEW

**Comment by shinezyy** [comment]
Can you provide the performance impact of this patch here?

**Comment by tastynoob** [comment]
image

**Comment by shinezyy** [comment]
> image

This is performance gap, but not the impact of this patch.
The impact of this patch means comparison between two GEM5 versions

**Comment by tastynoob** [comment]
image
disable prefetcher compare

File: src/cpu/o3/issue_queue.cc [hunk 1]
```diff
@@ -33,17 +32,31 @@
...
+#define RF_GETPRI(x) ((x) & 0b11)
+#define RF_GETPID(x) (((x) >> 2) & 0b1111)
+#define RF_GETTID(x) ((x) >> 6)
+
+#define RFMAKE_TPI(t, p) (((t) << 4) | (p))
+
+#define RF_INTID 0
+#define RF_FPID 1
+
+// rftpi = regfile typeid + portid
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: tpi seems to be strange.

TyPoID? or full name: typePortID

File: src/cpu/o3/issue_queue.cc [hunk 2]
```diff
@@ -241,7 +284,21 @@ IssueQue::retryMem(const DynInstPtr& inst)
     iqstats->retryMem++;
     DPRINTF(Schedule, "retry %s [sn:%llu]\n", enums::OpClassStrings[inst->opClass()], inst->seqNum);
     // scheduler->loadCancel(inst);
-    scheduler->addToFU(inst);
+    // scheduler->addToFU(inst);
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: remove useless code

File: src/cpu/o3/issue_queue.cc [hunk 3]
```diff
@@ -405,6 +474,7 @@ bool
 IssueQue::full()
 {
     bool full = instNumInsert + instNum >= iqsize;
+    full |= replayQ.size() > 32;  // TODO: parameterize it
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: Plz finish todo here

File: src/cpu/o3/issue_queue.cc [hunk 4]
```diff
@@ -539,27 +605,21 @@ Scheduler::SpecWakeupCompletion::description() const
...
-Scheduler::Scheduler(const SchedulerParams& params)
-    : SimObject(params), issueQues(params.IQs), intSlotNum(params.intSlotNum), fpSlotNum(params.fpSlotNum)
+Scheduler::Scheduler(const SchedulerParams& params) : SimObject(params), issueQues(params.IQs)
 {
     dispTable.resize(enums::OpClass::Num_OpClass);
     opExecTimeTable.resize(enums::OpClass::Num_OpClass, 1);
     opPipelined.resize(enums::OpClass::Num_OpClass, false);
 
     boost::dynamic_bitset<> opChecker(enums::Num_OpClass, 0);
+    std::vector<int> rfportChecker(64, 0);
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: The meaning of 64? add comments or replace it with constants

File: src/cpu/o3/issue_queue.cc [hunk 5]
```diff
@@ -573,7 +633,24 @@ Scheduler::Scheduler(const SchedulerParams& params)
...
+        }
+        for (auto rftpi : issueQues[i]->fpRfTPI) {
+            for (auto tpi : rftpi) {
+                maxTPI = std::max(maxTPI, tpi.first);
+                rfportChecker[tpi.first] += 1;
+            }
+        }
     }
+    maxTPI += 1;
+    assert(maxTPI <= 64);
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc]: The meaning of 64? add comments or replace it with constants

File: src/cpu/o3/issue_queue.cc [hunk 6]
```diff
@@ -573,7 +633,24 @@ Scheduler::Scheduler(const SchedulerParams& params)
                 opChecker.set(op->opClass);
             }
         }
+
+        for (auto rftpi : issueQues[i]->intRfTPI) {
+            for (auto tpi : rftpi) {
+                maxTPI = std::max(maxTPI, tpi.first);
+                rfportChecker[tpi.first] += 1;
+            }
+        }
```
Comment (shinezyy) [inline src/cpu/o3/issue_queue.cc:639]: Can we use `const auto &` here?

**Review by shinezyy — CHANGES_REQUESTED** [review]

File: src/cpu/o3/issue_queue.cc (hunk 1 above)
Comment (tastynoob) [reply src/cpu/o3/issue_queue.cc]: how about "TyPI"

File: src/cpu/o3/issue_queue.cc (hunk 3 above)
Comment (tastynoob) [reply src/cpu/o3/issue_queue.cc]: Maybe remove this future

**Review by shinezyy — APPROVED** [review]

**Pull request merged**